packages/esbuild/src/executors/esbuild/lib/build-esbuild-options.ts

```diff
@@ -123,6 +127,50 @@ export function buildEsbuildOptions(
   return esbuildOptions;
 }
 
+/**
+ * When using TS project references we need to map the paths to the referenced projects.
+ * This is necessary because esbuild does not support project references out of the box.
+ * @param context ExecutorContext
+ */
+export function createPathsFromTsConfigReferences(
+  context: ExecutorContext
+): Record<string, string[]> {
+  const {
+    findAllProjectNodeDependencies,
+  } = require('nx/src/utils/project-graph-utils');
+
+  const deps = findAllProjectNodeDependencies(context.projectName);
+  const tsConfig = readJsonFile(
+    joinPathFragments(context.root, 'tsconfig.json')
+  );
+  const referencesAsPaths = new Set(
+    tsConfig.references.map((ref) => joinPathFragments(workspaceRoot, ref.path))
+  );
+
+  // for each dep we check if it contains a build target
+  // we only want to add the paths for projects that do not have a build target
+  return deps.reduce((acc, dep) => {
+    const projectNode = context.projectGraph.nodes[dep];
+    if (!projectNode.data.targets?.build) {
+      const projectPath = joinPathFragments(
+        workspaceRoot,
+        projectNode.data.root
+      );
+      const projectPkgJson = readJsonFile(
+        joinPathFragments(projectPath, 'package.json')
+      );
+      // if the project has a name and a main entry point, we add it to the paths
+      if (projectPkgJson?.name && projectPkgJson?.main) {
```

Comment:
We must consider `exports` first, and only if `exports` is not defined can we fall back to `main` and/or `module.`